[6.1] Added Exception Message to Taks Notification Email

brbrbr opened the pull request:
### Summary of Changes

The email sent by the system plugin task notification contains very little useful information. This adjustment adds the Exception message to the emails.
The emails point to the  Joomla! backend for more information, however there is no info in the dashboard since the taskcontent is not stored and the logs are well hidden.

### Testing Instructions
Enable email notifications for failure for any active task for example in sessiongc. Is enabled by default.

Throw an exception in that task, for example:
`  private function sessionGC(ExecuteTaskEvent $event): int
    {
      
        throw new \Exception('test');
...
`
execute the task

### Actual result BEFORE applying this Pull Request

Email with little information

### Expected result AFTER applying this Pull Request

Email with the exception message



### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

brbrbr committed:
Added Exception Message to Taks Notification Email

brbrbr committed:
Merge branch 'joomla:4.4-dev' into tasknotification-with-exception

brianteeman commented:
If the email now includes the message is there any point in telling someone to go to the joomla admin etc?

brbrbr committed:
Removed reference to Joomla! admin.

@brianteeman in https://github.com/joomla/joomla-cms/issues/43666#issuecomment-2174438686_

brbrbr commented:
> If the email now includes the message is there any point in telling someone to go to the joomla admin etc?

Good point,

Actually, in the current situation, there is also no point of telling someone to go to the admin. Since there is no information there. 
(unless you have an extension to view the logs). 


Should I include the debug trace in the email as well? That one is neither stored nor logged.


Bram

brbrbr committed:
New Line

brbrbr committed:
Hello

brbrbr committed:
Added trace string

This information in lost after task completion

brbrbr committed:
Removed a merged file from pull request

brbrbr committed:
Merge branch 'tasknotification-with-exception' of https://github.com/brbrbr/joomla-cms into tasknotification-with-exception

brbrbr committed:
php-cs-fixer

brbrbr committed:
Typoes

richard67 commented on plugins/system/tasknotification/src/Extension/TaskNotification.php:
```diff
@@ -241,6 +241,10 @@ private function getDataFromTask(Task $task): array
+            'TASK_EXCEPTION' => ($task->getContent()['exception']?->getMessage() ?? '')
+                    ."\n\n". // This might result in one or two empty strings with new Lines. Not too visible in email or HTML, so not really worth the effort to add a lot of extra code to optimize this.
+                 Path::removeRoot($task->getContent()['exception']?->getTraceAsString() ?? '')
+                ,
```
```suggestion
            'TASK_EXCEPTION' => ($task->getContent()['exception']?->getMessage() ?? '')
                . "\n\n" // This might result in one or two empty strings with new Lines. Not too visible in email or HTML, so not really worth the effort to add a lot of extra code to optimize this.
                . Path::removeRoot($task->getContent()['exception']?->getTraceAsString() ?? ''),
```
Fix code style check (PHPCS) failing in CI (drone).

brbrbr committed:
phpcs fix

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

Quy commented on administrator/language/en-GB/plg_system_tasknotification.ini:
```diff
@@ -4,9 +4,9 @@
 
 PLG_SYSTEM_TASK_NOTIFICATION="System - Task Notification"
-PLG_SYSTEM_TASK_NOTIFICATION_FAILURE_MAIL_BODY="Hello,\n\n\nPlanned execution of Scheduled Task#{TASK_ID}, {TASK_TITLE}, has failed with exit code {EXIT_CODE} at {EXEC_DATE_TIME}.\n\nPlease visit the Joomla! backend for more information.\n\n{TASK_OUTPUT}"
+PLG_SYSTEM_TASK_NOTIFICATION_FAILURE_MAIL_BODY="Hello,\n\nExecution of Scheduled Task#{TASK_ID}, {TASK_TITLE}, has failed with exit code {EXIT_CODE} at {EXEC_DATE_TIME}. The error message is shown below. The problem might be temporarily or related to this scheduled task only. However, it might impact your site as well. Please verify that your site is working correctly as soon as possible..\n{TASK ...
```
```suggestion
PLG_SYSTEM_TASK_NOTIFICATION_FAILURE_MAIL_BODY="Hello,\n\nExecution of Scheduled Task#{TASK_ID}, {TASK_TITLE}, has failed with exit code {EXIT_CODE} at {EXEC_DATE_TIME}. The error message is shown below. The problem might be temporarily or related to this scheduled task only. However, it might impact your site as well. Please verify that your site is working correctly as soon as possible.\n{TASK_OUTPUT}\nError Message (if any):\n{TASK_EXCEPTION}"
```

Quy commented on administrator/language/en-GB/plg_system_tasknotification.ini:
```diff
@@ -4,9 +4,9 @@
+PLG_SYSTEM_TASK_NOTIFICATION_FAILURE_MAIL_BODY="Hello,\n\nExecution of Scheduled Task#{TASK_ID}, {TASK_TITLE}, has failed with exit code {EXIT_CODE} at {EXEC_DATE_TIME}. The error message is shown below. The problem might be temporarily or related to this scheduled task only. However, it might impact your site as well. Please verify that your site is working correctly as soon as possible..\n{TASK ...
 PLG_SYSTEM_TASK_NOTIFICATION_FAILURE_MAIL_SUBJECT="Task Failure"
-PLG_SYSTEM_TASK_NOTIFICATION_FATAL_MAIL_BODY="Hello,\n\nPlanned execution of Scheduler Task#{TASK_ID}, {TASK_TITLE}, recovered from a fatal failure.\n\nThis could mean that the task execution exhausted the system resources or the restrictions from the PHP INI.\n\nPlease visit the Joomla! backend for more information."
+PLG_SYSTEM_TASK_NOTIFICATION_FATAL_MAIL_BODY="Hello,\n\nJoomla! caught a fatal failure while executing the Scheduled Task#{TASK_ID}, {TASK_TITLE}  at {EXEC_DATE_TIME}. The error message is shown below. This could mean that the task execution exhausted the system resources or the restrictions from the PHP INI. The problem might be temporarily or related to this scheduled task only. However, it mig ...
```
```suggestion
PLG_SYSTEM_TASK_NOTIFICATION_FATAL_MAIL_BODY="Hello,\n\nJoomla! caught a fatal failure while executing the Scheduled Task#{TASK_ID}, {TASK_TITLE}  at {EXEC_DATE_TIME}. The error message is shown below. This could mean that the task execution exhausted the system resources or the restrictions from the PHP INI. The problem might be temporarily or related to this scheduled task only. However, it might impact your site as well. Please verify that your site is working correctly as soon as possible.\n{TASK_OUTPUT}\nError Message (if any):\n{TASK_EXCEPTION}"
```

brbrbr committed:
Update administrator/language/en-GB/plg_system_tasknotification.ini

Co-authored-by: Quy <quy@nomonkeybiz.com>

brbrbr committed:
Update administrator/language/en-GB/plg_system_tasknotification.ini

Co-authored-by: Quy <quy@nomonkeybiz.com>

brbrbr committed:
Merge branch '5.2-dev' into tasknotification-with-exception

HLeithner commented:
This pull request has been automatically rebased to 5.3-dev.

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

HLeithner commented:
This pull request has been automatically rebased to 6.1-dev.
